content/issue-2/applicative-parsing.md

```diff
@@ -0,0 +1,711 @@
+
+At this point parser can represent arguments of any type and provides a way to map between types
+
+## Handling failures
+
+Supporting mapping that can't fail is easy with just the `Functor` abstraction but it can't
+handle failures well - consider parsing numeric information:
+
+```
+$ app --answer 42
+$ app --answer Forty-two
+```
+
+A parser using `map`, `FromStr::from_str` and `unwrap` can parse the first line but panics
+with a bad error message for the second one. To handle parsing failures `Magic` needs to be
+able to represent them. The usual approach for dealing with failing computations in Rust is
+with help from the `Result` type.
+
+```rust
+struct Magic<T>(Result<T, String>);
+impl<T> Magic<T> {
+    fn run(self) -> Result<T, String> {
+        self.0
+    }
+
+    fn parse<R, F: Fn(T) -> Result<R, String>>(self, f: F) -> Magic<R> {
+        match self.0 {
+            Ok(x) => Magic(f(x)),
+            Err(err) => Magic(Err(err)),
+        }
+    }
+```
+
+Function `parse` applies a failing computation on a value inside `Magic` if one exists or keeps
+the error message untouched otherwise. Change in the representation requires to change `map` and
+`parse` can help with that too.
+
+```rust
+impl<T> Magic<T> {
+    fn map<R, F: Fn(T) -> R>(self, f: F) -> Magic<R> {
+        self.parse(|x| Ok(f(x)))
+    }
+}
+
+#[test]
+fn sample() {
+    use std::str::FromStr;
+    // --answer 42
+    let answer = Magic(Ok("42"));
+    let _result = answer
+        .parse(|n| u32::from_str(n).map_err(|e| e.to_string()))
+        .run(); // Ok(42)
+    todo!("{:?}", _result);
+
+    // --answer Forty-two => Err("invalid digit found in string")
+}
+```
+
+At this point the parser can represent arguments of any type and failures too. Function `parse` is an
+ad hoc thing and isn't coming from Category Theory.
+
+## Composing failing computations
+
+`run` makes it easy to check if computation is successful and to proceed only when it is, but it
+also adds a new corner case: consider an app that takes a name and the answer, checks
+if the answer is correct and reports to the user:
+
+```rust
+let name = match name.run() {
+    Ok(ok) => ok,
+    _ => panic!("You need to specify name"),
+}
+
+// println!("Hello {name}!") // (1)
+
+let answer = match answer.run().expect("You need to specify the answer");
+
+// println!("Hello {name}!") // (2)
+
+if answer == 42 {
+    println!("Your answer is correct");
+} else {
+    println!("Your answer is wrong");
+}
+```
+
+Greeting code can go in one of two position: (1) and (2), in the first position it executes
+before all the arguments are validated. In this example failure to validate all the
+arguments results in a confusing messages to user, but it's easy imagine a situation where an
+app might perform some harder to undo actions. Because of this a good argument parser needs to
+have a way to make sure all the arguments are validated before proceeding.
```

Comment:
```suggestion
before all the arguments are validated. In this example a failure to validate the
first argument results in a confusing message to user, but it's easy imagine a situation where, instead of writing a message, an
app might perform some harder to undo actions. Because of this, a good argument parser needs to
have a way to make sure all the arguments are validated before proceeding.
```